File: common/cpp/robocin/network/zmq_publisher_socket.cpp
#include "robocin/network/zmq_publisher_socket.h"

namespace vision {

ZmqPublisherSocket::ZmqPublisherSocket(int n_threads) :
    context_(n_threads),
    socket_(context_, ZMQ_PUB) {}

void ZmqPublisherSocket::bind(std::string_view address) { socket_.bind(std::string{address}); }

void ZmqPublisherSocket::send(std::string_view topic, std::string_view message) { // NOLINT
  zmq::message_t zmq_topic(topic);
  socket_.send(zmq_topic, zmq::send_flags::sndmore);

  zmq::message_t zmq_message(message);
  socket_.send(zmq_message, zmq::send_flags::dontwait);
}

void ZmqPublisherSocket::close() { zmq_close(socket_); }

} // namespace vision

File: tester/network/zmq_publisher_socket.py
from typing import Union
from network.pub_sub_mode import PubSubMode
from google.protobuf.message import Message as Protobuf
import zmq


class ZmqPublisherSocket:
    def __init__(self, n_threads: int = 1, timeout_ms: int = 3):
        # Create a ZMQ Publisher
        self.context = zmq.Context(n_threads)
        self.timeout_ms = timeout_ms
        self.publisher = self.context.socket(zmq.PUB)

    def bind(self, address):
        # Bind the socket to the given ZMQ address
        self.publisher.bind(address)
        self.publisher.setsockopt(zmq.SNDTIMEO, self.timeout_ms)

    def send(
        self,
        topic: str,
        message: Union[str, bytes, Protobuf],
        mode: PubSubMode = PubSubMode.DontWait,
    ):
        flags = zmq.DONTWAIT if mode == PubSubMode.DontWait else zmq.Flag(0)

        if isinstance(message, str):
            self.publisher.send_multipart(
                [topic.encode(), message.encode()], flags=flags
            )
        elif isinstance(message, bytes):
            self.publisher.send_multipart([topic.encode(), message], flags=flags)
        elif isinstance(message, Protobuf):
            self.publisher.send_multipart(
                [topic.encode(), message.SerializeToString()], flags=flags
            )
        else:
            raise ValueError(f"Invalid message type: {type(message)}.")
